src/server/scripts/Outland/zone_shadowmoon_valley.cpp

```diff
@@ -571,64 +589,105 @@ class npc_dragonmaw_peon : public CreatureScript
-            {
-                PlayerGUID = caster->GetGUID();
+            PlayerGUID = caster->GetGUID();
 
-                Tapped = true;
-                float x, y, z;
+            if (caster->GetTypeId() == TYPEID_PLAYER && spell->Id == SPELL_POISON)
+            {
                 caster->GetClosePoint(x, y, z, me->GetObjectSize());
-
-                me->SetWalk(false);
-                me->GetMotionMaster()->MovePoint(1, x, y, z);
+                Talk(SAY_1);
+                events2.ScheduleEvent(EVENT_WALK_TO_MUTTON, 0);
             }
         }
 
-        void MovementInform(uint32 type, uint32 id)
+        void MovementInform(uint32 /*type*/, uint32 id)
         {
-            if (type != POINT_MOTION_TYPE)
-                return;
+            if (id == 1)
+            {
+                me->HandleEmoteCommand(EMOTE_ONESHOT_EAT);
+                events2.ScheduleEvent(EVENT_POISIONED, 5000);
+            }
+        }
 
-            if (id)
+        void CreditPlayer()
+        {
+            if (PlayerGUID)
             {
-                me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_ONESHOT_EAT);
-                PoisonTimer = 15000;
+                Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID);
+                if (player && player->GetQuestStatus(11020) == QUEST_STATUS_INCOMPLETE)
+                    player->KilledMonsterCredit(23209, 0);
             }
         }
 
         void UpdateAI(uint32 diff)
         {
-            if (PoisonTimer)
+            events.Update(diff);
+            events2.Update(diff);
+
+            if (!UpdateVictim())
             {
-                if (PoisonTimer <= diff)
+                switch (events2.ExecuteEvent())
                 {
-                    if (PlayerGUID)
-                    {
-                        Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID);
-                        if (player && player->GetQuestStatus(11020) == QUEST_STATUS_INCOMPLETE)
-                            player->KilledMonsterCredit(23209, 0);
-                    }
-                    PoisonTimer = 0;
+
+                case EVENT_WALK_TO_MUTTON:
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(1, x, y, z, true);
+                    me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_STATE_NONE);
+                    me->HandleEmoteCommand(EMOTE_ONESHOT_CHEER);
+                    break;
+                case EVENT_POISIONED:
+                    if (roll_chance_i(20))
+                        Talk(SAY_POISONED_1);
+                    me->CastSpell(me, SPELL_VOMIT);
+                    events2.ScheduleEvent(EVENT_KILL, 5000);
+                    break;
+                case EVENT_KILL:
                     Unit::DealDamage(me, me, me->GetHealth(), NULL, DIRECT_DAMAGE, SPELL_SCHOOL_MASK_NORMAL, NULL, false);
-                } else PoisonTimer -= diff;
+                    CreditPlayer();
```

Comment:
- CreditPlayer();
+ if (GameObject* food = me->FindNearestGameObject(185893, 5.0f))
+    food->RemoveFromWorld();

Reply:
Why?

Reply:
NPC removes food from the side after poisoning